File: Assignment_1/Task_2.py
# Modified Python code to print the sorted array

import numpy as np
import time

# Merge Sort implementation
def merge_sort(arr):
    if len(arr) > 1:
        mid = len(arr) // 2
        L = arr[:mid]
        R = arr[mid:]
        
        merge_sort(L)
        merge_sort(R)
        
        i = j = k = 0
        
        while i < len(L) and j < len(R):
            if L[i] < R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1

        while i < len(L):
            arr[k] = L[i]
            i += 1
            k += 1

        while j < len(R):
            arr[k] = R[j]
            j += 1
            k += 1

# Quick Sort implementation
def quick_sort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)
        quick_sort(arr, low, pi - 1)
        quick_sort(arr, pi + 1, high)

def partition(arr, low, high):
    pivot = arr[high]
    i = low - 1
    for j in range(low, high):
        if arr[j] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

# Function to measure execution time
def measure_execution_time(sort_function, arr):
    start_time = time.time()
    if sort_function == quick_sort:
        sort_function(arr, 0, len(arr) - 1)
    else:
        sort_function(arr)
    end_time = time.time()
    return (end_time - start_time) * 1000  # Convert to milliseconds

# Function to print an array
def print_array(arr, name):
    print(f"{name} Sorted Array:")
    print(arr[:100])  # Print only the first 100 elements for readability

# Using the same input size as in the C program (n = 10000)
n = 10000
arr = np.random.randint(0, 10000, size=n).tolist()

# Measure execution time for Merge Sort
merge_arr = arr.copy()
print("Original Array (First 100 Elements):")
print(merge_arr[:100])  # Print a sample of the original array

merge_time = measure_execution_time(merge_sort, merge_arr)
print_array(merge_arr, "Merge Sort")

# Measure execution time for Quick Sort
quick_arr = arr.copy()
quick_time = measure_execution_time(quick_sort, quick_arr)
print_array(quick_arr, "Quick Sort")

# Display execution times
print(f"\nExecution Time for Merge Sort: {merge_time:.3f} ms")
print(f"Execution Time for Quick Sort: {quick_time:.3f} ms")

File: Assignment_1/Task_2.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Function to print an array
void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

// Merge Sort Implementation
void merge(int arr[], int left, int mid, int right) {
    int i, j, k;
    int n1 = mid - left + 1;
    int n2 = right - mid;
    int L[n1], R[n2];

    for (i = 0; i < n1; i++)
        L[i] = arr[left + i];
    for (j = 0; j < n2; j++)
        R[j] = arr[mid + 1 + j];

    i = 0;
    j = 0;
    k = left;
    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k] = L[i];
            i++;
        } else {
            arr[k] = R[j];
            j++;
        }
        k++;
    }

    while (i < n1) {
        arr[k] = L[i];
        i++;
        k++;
    }
    while (j < n2) {
        arr[k] = R[j];
        j++;
        k++;
    }
}

void mergeSort(int arr[], int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }
}

// Quick Sort Implementation
int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = (low - 1);
    for (int j = low; j < high; j++) {
        if (arr[j] < pivot) {
            i++;
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }
    int temp = arr[i + 1];
    arr[i + 1] = arr[high];
    arr[high] = temp;
    return (i + 1);
}

void quickSort(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

// Function to measure execution time
void measureExecutionTime(void (*sortFunction)(int[], int, int), int arr[], int n) {
    clock_t start, end;
    start = clock();
    sortFunction(arr, 0, n - 1);
    end = clock();
    printf("Execution time: %f ms\n", ((double)(end - start) / CLOCKS_PER_SEC) * 1000);
}

int main() {
    int n = 10000;
    int *arr = (int *)malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) {
        arr[i] = rand() % 10000;
    }

    printf("Original Array (First 100 Elements):\n");
    printArray(arr, 100);

    printf("\nMerge Sort:\n");
    measureExecutionTime(mergeSort, arr, n);
    printf("Sorted Array (First 100 Elements):\n");
    printArray(arr, 100);
    
    for (int i = 0; i < n; i++) {
        arr[i] = rand() % 10000;
    }

    printf("\nOriginal Array (First 100 Elements):\n");
    printArray(arr, 100);
    
    printf("\nQuick Sort:\n");
    measureExecutionTime(quickSort, arr, n);
    printf("Sorted Array (First 100 Elements):\n");
    printArray(arr, 100);
    
    free(arr);
    return 0;
}
